### ISSUE TITLE
Multiple CMake target errors ever since commit 0e52627358

### ISSUE THREAD

**Issue by dncliss** [pytorch/pytorch#34606]
## 🐛 Bug

Building pytorch on platform `ppc64le`, since Feb 28, has resulted in a flurry of CMake Error
messages partway through the build.  An example of the error message appears below, though the error will repeat with some variations and different target names:

```core header install: /home/jenkins/pytorch/build/aten/src/ATen/core/TensorBody.h
core header install: /home/jenkins/pytorch/build/aten/src/ATen/core/TensorMethods.h
CMake Error at cmake/Codegen.cmake:207 (add_custom_target):
  add_custom_target cannot create target "ATEN_CPU_FILES_GEN_TARGET" because
  another target with the same name already exists.  The existing target is a
  custom target created in source directory "/home/jenkins/pytorch/caffe2".
  See documentation for policy CMP0002 for more details.
Call Stack (most recent call first):
  caffe2/CMakeLists.txt:2 (include)
```

To see a full log with the varying CMake errors, here is an example CI output on ppc64le:
https://powerci.osuosl.org/job/pytorch-linux-cuda92-cudnn7-py3-mpi-build-only-gpu/518/consoleFull

## To Reproduce

1. Run the build scripts for pytorch on a `ppc64le` system.  In my case, it is a docker container
running ubuntu and using cuda10.2 (or 10.1 or 10.0) but i'm doubtful the OS or CUDA level matters.  Wait a few minutes; the errors seem to start occuring in under 5 min.

NOTE:
I manually tried different builds with different levels of pytorch master until I narrowed it
down to the exact pull request on which the problem started to happen:  The commit ID
where it starts failing is `0e52627358` -- this corresponds to PR #338669.  The errors
don't appear if I simply build from the prior commit.  
Reference:  https://github.com/pytorch/pytorch/commit/0e52627358c4ec60243f9d8e19b722b7b2265ffc

Not being an expert in this, I haven't determined what it is in this commit that triggered the error,
but it has happened continuously in every CI run since that change was included.
I'll continue to look a bit closer, but I wanted to get the issue documented.  Perhaps @kimishpatel has some suggestions or thoughts on what to look at (as user who made the aforementioned commit)?

## Environment

Ubuntu 18.04.4
CUDA 10.2 (also reproduces on 10.1 and 10.0).
Build using current master branch of pytorch, or any since 0e52627358
Built on ppc64le in a docker container, but I don't know if this is necessarily tied to
the one platform or not.

**Comment by dncliss** [comment]
Main page with nightly CI builds on `ppc64le`; you can see the failures every night starting Feb 29.  Output is available to browse for any of these runs, though the failure messages are all
essentially the same since that date:
https://powerci.osuosl.org/job/pytorch-linux-cuda92-cudnn7-py3-mpi-build-only-gpu/

**Comment by kimishpatel** [comment]
Thank you @dncliss for bringing this up. It may need a little more digging in to figure out what is happening. From the looks of it is trying to redefine ATEN_CPU_FILES_GEN_TARGET target, but nothing seems obvious why this should be. I will try to take a look, but it may take some time.

**Comment by kimishpatel** [comment]
@dncliss, maybe a quick way to check if that commit is a problem create a trial branch and apply revert of that commit on it?

**Comment by dncliss** [comment]
@kimishpatel - thanks for your thoughts on this.  I'll keep looking myself.  As for the ATEN_CPU_FILES_GEN_TARGET, that's just one of many, so the issue isn't specific to that target; there's something else underlying that is having an impact.

To find the specific commit that introduced the issue, I looked for likely commit targets on Feb 28/29 when the nightly builds started failing.  I looked for candidates (e.g. commits with CMakefile changes) and tried building at a few different commit points based on the history:
https://github.com/pytorch/pytorch/commits/master
I observed the error when building at 0e52627358; I then tried building from the prior commit (85b1c45) and it succeeded.  I reset back up to 0e52627358 and repeated, with the failure returning.
So .....  I don't have a great understanding of the code myself, but I will review it some more and see if I can figure anything out.  If you get a chance to revisit/review this PR (since you're more familiar with it, having made the change) and have any ideas, let me know.  If I learn anything new myself I'll also post back here.

**Comment by dncliss** [comment]
Updating: Experimenting a little more today:

```
0e52627358 (HEAD) Fixing pthreadpool symbol conflict issue. (#33869)
85b1c45a45 (origin/gh/zasdfgbnm/20/base) [JIT] fix alias assertion (#33952)

BUILD FROM 0e52627358:   Failure (as documented above)
BUILD FROM 85b1c45a45:   Clean build

BUILD FROM 0e52627358, but:
 Revert both caffe2/CMakeList.txt and caffe2/utils/CMakeList.txt to 85b1c45a45
     ---> Clean build
 Revert only caffe2/utils/CMakeList.txt to 85b1c45a45
     ---> Failure (as above)
 Revert only caffe2/CMakeList.txt to 85b1c45a45
     ---> Different build failures
```

The last scenario above is probably not worth looking at, being a bogus mixture of
two commits and it demonstrating the ongoing error in question.

The rest of 0e52627358 doesn't cause breakage without the 2 corresponding CMakeList.txt
file changes in place.
The failure *is* there with new caffe2/CMakeList.txt and old caffe2/utils/CMakeList.txt,
so a change in the latter isn't injecting the issue (since it's observed even using the
prior version).   To me, that means we can just look at caffe2/CMakeList.txt  ..... (tbd)

**Comment by kimishpatel** [comment]
Try removing these lines:   https://github.com/pytorch/pytorch/commit/0e52627358c4ec60243f9d8e19b722b7b2265ffc#diff-8f885316d57c59b5aa82e9857dbe5935R62-R65?
I dont suppose "-fno-openmp" flag applies to your build.
If not that it might be related to this: https://github.com/pytorch/pytorch/commit/0e52627358c4ec60243f9d8e19b722b7b2265ffc#diff-8f885316d57c59b5aa82e9857dbe5935R55-R61
Unfortunately fxdiv is a dependency for the files added in that commit. But if fxdiv turns out to be the issue, we can try to figure out what to do.

**Comment by dncliss** [comment]
Yep....narrowing down: removing lines 62-65, the error remains as described.  (Also, removing 695-697 only, same thing.)
But keeping those and removing only lines 55-61, the fxdiv stuff, makes the compile choke - demonstrating as you say, that this is a fundamental element of that commit.  
I'm wondering also why (or if?) this is only failing on our PPC64LE platform for some reason.  But even if so, it's not like we can just add a platform condition check here, since omitting these lines means failure.  (For grins, I tried omitting just 58-60, but no; those are essential.)

FYI; I'll be mostly out of pocket until next Wed (18th), so barring any new inspiration, this issue may stay on hold for a bit.

**Comment by dncliss** [comment]
@kimishpatel - I'm back to looking at this again.  So this does appear tied to the `fxdiv` code additions. Presumably in spite of the failures on PPC64LE it must be working on X86, or the issue would have been raised sooner (?)

One theory, considering what is different between the builds on these two platforms:  On PPC, we don't have the QNNPACK support and the USE_QNNPACK boolean check is thus off for our build.  Does the `fxdiv` code expect QNNPACK to be present?  Is the code possibly something that should have been qualified within an "if (USE_QNNPACK) ..." check?
(It's only a hypothesis, as reason for the possible difference; I admit I am coming from a position of relative ignorance on the topic.)

**Comment by rjknight** [comment]
I ended up hitting this issue trying to compile the v1.5 branch also, I followed along with @dncliss 's suggestion that maybe some of the additions needed to be qualified with the new USE_XNNPACK adding the following changes seem to get me past the cmake failures
```
diff --git a/caffe2/CMakeLists.txt b/caffe2/CMakeLists.txt
index 7a59bd5..fd5dbf9 100644
--- a/caffe2/CMakeLists.txt
+++ b/caffe2/CMakeLists.txt
@@ -51,18 +51,18 @@ endif()
 # addressed yet.
 
 # For pthreadpool_new_if_impl. TODO: Remove when threadpools are unitied.
-if (NOT MSVC)
-  IF(NOT TARGET fxdiv)
-    SET(FXDIV_BUILD_TESTS OFF CACHE BOOL "")
-    SET(FXDIV_BUILD_BENCHMARKS OFF CACHE BOOL "")
-    ADD_SUBDIRECTORY(
-      "${FXDIV_SOURCE_DIR}"
-      "${CMAKE_BINARY_DIR}/FXdiv")
-  ENDIF()
-  if (NOT (INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE))
-    set_source_files_properties(
-        utils/threadpool/pthreadpool_new_if_impl.c PROPERTIES COMPILE_FLAGS -fno-openmp)
-  endif()
+if (NOT MSVC AND USE_XNNPACK)
+ IF(NOT TARGET fxdiv)
+   SET(FXDIV_BUILD_TESTS OFF CACHE BOOL "")
+   SET(FXDIV_BUILD_BENCHMARKS OFF CACHE BOOL "")
+   ADD_SUBDIRECTORY(
+     "${FXDIV_SOURCE_DIR}"
+     "${CMAKE_BINARY_DIR}/FXdiv")
+ ENDIF()
+ if (NOT (INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE))
+   set_source_files_properties(
+       utils/threadpool/pthreadpool_new_if_impl.c PROPERTIES COMPILE_FLAGS -fno-openmp)
+ endif()
 endif()
 
 add_subdirectory(core)
@@ -710,7 +710,7 @@ ELSEIF(USE_CUDA)
 ENDIF()
 
 
-if (NOT MSVC)
+if (NOT MSVC AND USE_XNNPACK)
   TARGET_LINK_LIBRARIES(torch_cpu PRIVATE fxdiv)
 endif()
 
diff --git a/caffe2/utils/CMakeLists.txt b/caffe2/utils/CMakeLists.txt
index 27aabb1..e9792d2 100644
--- a/caffe2/utils/CMakeLists.txt
+++ b/caffe2/utils/CMakeLists.txt
@@ -36,7 +36,7 @@ list(APPEND Caffe2_CPU_SRCS
 # ---[ threadpool/pthreadpool* is a local modification of the NNPACK
 # pthreadpool with a very similar interface. Neither NNPACK, nor this
 # thread pool supports Windows.
-if (NOT MSVC)
+if (NOT MSVC AND USE_XNNPACK)
   add_definitions(-DUSE_INTERNAL_THREADPOOL_IMPL)
   set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS}
           utils/threadpool/pthreadpool.cc

```
Comparing the difference between x86 and ppc64le builds, I found that on my ppc64le system FXDIV_SOURCE_DIR was empty in the path above, while on my  x86 system the value was always set
 
Then in the log I noticed this warning...
```
 386 CMake Warning at cmake/Dependencies.cmake:220 (message):
 387   Target architecture "ppc64le" is not supported in {Q/X}NNPACK.  Supported
 388   architectures are x86, x86-64, ARM, and ARM64.  Turn this warning off by
 389   USE_{Q/X}NNPACK=OFF.
```
Looking at that code path
```
 219     if (NOT IOS AND NOT (CMAKE_SYSTEM_PROCESSOR MATCHES "^(i686|AMD64|x86_64|armv[0-9].*|arm64|aarch64)$"))
 220       message(WARNING
 221         "Target architecture \"${CMAKE_SYSTEM_PROCESSOR}\" is not supported in {Q/X}NNPACK. "
 222         "Supported architectures are x86, x86-64, ARM, and ARM64. "
 223         "Turn this warning off by USE_{Q/X}NNPACK=OFF.")
 224       set(DISABLE_NNPACK_AND_FAMILY ON)
 225     endif()
 226   endif()
 227 
 228   if (DISABLE_NNPACK_AND_FAMILY)
 229     set(USE_NNPACK OFF)
 230     set(USE_QNNPACK OFF)
 231     set(USE_PYTORCH_QNNPACK OFF)
 232     set(USE_XNNPACK OFF)
 233   else()
 234     set(CAFFE2_THIRD_PARTY_ROOT "${PROJECT_SOURCE_DIR}/third_party")
 235 
 236     if (NOT DEFINED CPUINFO_SOURCE_DIR)
 237       set(CPUINFO_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/cpuinfo" CACHE STRING "cpuinfo source directory")
 238     endif()
 239     if (NOT DEFINED FP16_SOURCE_DIR)
 240       set(FP16_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/FP16" CACHE STRING "FP16 source directory")
 241     endif()
 242     if (NOT DEFINED FXDIV_SOURCE_DIR)
 243       set(FXDIV_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/FXdiv" CACHE STRING "FXdiv source directory")
 244     endif()
 245     if (NOT DEFINED PSIMD_SOURCE_DIR)
 246       set(PSIMD_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/psimd" CACHE STRING "PSimd source directory")
 247     endif()
 248     if (NOT DEFINED PTHREADPOOL_SOURCE_DIR)
 249       set(PTHREADPOOL_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/pthreadpool" CACHE STRING "pthreadpool source directory")
 250     endif()
 251 
 252     set(CPUINFO_LIBRARY_TYPE "static" CACHE STRING "")
 253     set(CPUINFO_LOG_LEVEL "error" CACHE STRING "")
 254     set(PTHREADPOOL_LIBRARY_TYPE "static" CACHE STRING "")
 255   endif()
 256 endif()
```
we can see that FXDIV_SOURCE_PATH does not get set if the warning is displayed, and the code disables all the xNNPACK stuff.

I tested the theory by forcing CMake to go through the fail path on my x86 system, resulting in the FDIV_SOURCE_PATH being empty and saw the same cmake errors in the build.

Currently, I have my code patched with the change I listed above and would be willing to put up a PR for the fix, but would need a little guidance as to whether there are other code paths which would need to be qualified to fully turn off this feature in the cases it is not supported or the user unselects it.

**Comment by cyyever** [comment]
> I ended up hitting this issue trying to compile the v1.5 branch also, I followed along with @dncliss 's suggestion that maybe some of the additions needed to be qualified with the new USE_XNNPACK adding the following changes seem to get me past the cmake failures
> 
> ```
> diff --git a/caffe2/CMakeLists.txt b/caffe2/CMakeLists.txt
> index 7a59bd5..fd5dbf9 100644
> --- a/caffe2/CMakeLists.txt
> +++ b/caffe2/CMakeLists.txt
> @@ -51,18 +51,18 @@ endif()
>  # addressed yet.
>  
>  # For pthreadpool_new_if_impl. TODO: Remove when threadpools are unitied.
> -if (NOT MSVC)
> -  IF(NOT TARGET fxdiv)
> -    SET(FXDIV_BUILD_TESTS OFF CACHE BOOL "")
> -    SET(FXDIV_BUILD_BENCHMARKS OFF CACHE BOOL "")
> -    ADD_SUBDIRECTORY(
> -      "${FXDIV_SOURCE_DIR}"
> -      "${CMAKE_BINARY_DIR}/FXdiv")
> -  ENDIF()
> -  if (NOT (INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE))
> -    set_source_files_properties(
> -        utils/threadpool/pthreadpool_new_if_impl.c PROPERTIES COMPILE_FLAGS -fno-openmp)
> -  endif()
> +if (NOT MSVC AND USE_XNNPACK)
> + IF(NOT TARGET fxdiv)
> +   SET(FXDIV_BUILD_TESTS OFF CACHE BOOL "")
> +   SET(FXDIV_BUILD_BENCHMARKS OFF CACHE BOOL "")
> +   ADD_SUBDIRECTORY(
> +     "${FXDIV_SOURCE_DIR}"
> +     "${CMAKE_BINARY_DIR}/FXdiv")
> + ENDIF()
> + if (NOT (INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE))
> +   set_source_files_properties(
> +       utils/threadpool/pthreadpool_new_if_impl.c PROPERTIES COMPILE_FLAGS -fno-openmp)
> + endif()
>  endif()
>  
>  add_subdirectory(core)
> @@ -710,7 +710,7 @@ ELSEIF(USE_CUDA)
>  ENDIF()
>  
>  
> -if (NOT MSVC)
> +if (NOT MSVC AND USE_XNNPACK)
>    TARGET_LINK_LIBRARIES(torch_cpu PRIVATE fxdiv)
>  endif()
>  
> diff --git a/caffe2/utils/CMakeLists.txt b/caffe2/utils/CMakeLists.txt
> index 27aabb1..e9792d2 100644
> --- a/caffe2/utils/CMakeLists.txt
> +++ b/caffe2/utils/CMakeLists.txt
> @@ -36,7 +36,7 @@ list(APPEND Caffe2_CPU_SRCS
>  # ---[ threadpool/pthreadpool* is a local modification of the NNPACK
>  # pthreadpool with a very similar interface. Neither NNPACK, nor this
>  # thread pool supports Windows.
> -if (NOT MSVC)
> +if (NOT MSVC AND USE_XNNPACK)
>    add_definitions(-DUSE_INTERNAL_THREADPOOL_IMPL)
>    set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS}
>            utils/threadpool/pthreadpool.cc
> ```
> 
> Comparing the difference between x86 and ppc64le builds, I found that on my ppc64le system FXDIV_SOURCE_DIR was empty in the path above, while on my x86 system the value was always set
> 
> Then in the log I noticed this warning...
> 
> ```
>  386 CMake Warning at cmake/Dependencies.cmake:220 (message):
>  387   Target architecture "ppc64le" is not supported in {Q/X}NNPACK.  Supported
>  388   architectures are x86, x86-64, ARM, and ARM64.  Turn this warning off by
>  389   USE_{Q/X}NNPACK=OFF.
> ```
> 
> Looking at that code path
> 
> ```
>  219     if (NOT IOS AND NOT (CMAKE_SYSTEM_PROCESSOR MATCHES "^(i686|AMD64|x86_64|armv[0-9].*|arm64|aarch64)$"))
>  220       message(WARNING
>  221         "Target architecture \"${CMAKE_SYSTEM_PROCESSOR}\" is not supported in {Q/X}NNPACK. "
>  222         "Supported architectures are x86, x86-64, ARM, and ARM64. "
>  223         "Turn this warning off by USE_{Q/X}NNPACK=OFF.")
>  224       set(DISABLE_NNPACK_AND_FAMILY ON)
>  225     endif()
>  226   endif()
>  227 
>  228   if (DISABLE_NNPACK_AND_FAMILY)
>  229     set(USE_NNPACK OFF)
>  230     set(USE_QNNPACK OFF)
>  231     set(USE_PYTORCH_QNNPACK OFF)
>  232     set(USE_XNNPACK OFF)
>  233   else()
>  234     set(CAFFE2_THIRD_PARTY_ROOT "${PROJECT_SOURCE_DIR}/third_party")
>  235 
>  236     if (NOT DEFINED CPUINFO_SOURCE_DIR)
>  237       set(CPUINFO_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/cpuinfo" CACHE STRING "cpuinfo source directory")
>  238     endif()
>  239     if (NOT DEFINED FP16_SOURCE_DIR)
>  240       set(FP16_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/FP16" CACHE STRING "FP16 source directory")
>  241     endif()
>  242     if (NOT DEFINED FXDIV_SOURCE_DIR)
>  243       set(FXDIV_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/FXdiv" CACHE STRING "FXdiv source directory")
>  244     endif()
>  245     if (NOT DEFINED PSIMD_SOURCE_DIR)
>  246       set(PSIMD_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/psimd" CACHE STRING "PSimd source directory")
>  247     endif()
>  248     if (NOT DEFINED PTHREADPOOL_SOURCE_DIR)
>  249       set(PTHREADPOOL_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/pthreadpool" CACHE STRING "pthreadpool source directory")
>  250     endif()
>  251 
>  252     set(CPUINFO_LIBRARY_TYPE "static" CACHE STRING "")
>  253     set(CPUINFO_LOG_LEVEL "error" CACHE STRING "")
>  254     set(PTHREADPOOL_LIBRARY_TYPE "static" CACHE STRING "")
>  255   endif()
>  256 endif()
> ```
> 
> we can see that FXDIV_SOURCE_PATH does not get set if the warning is displayed, and the code disables all the xNNPACK stuff.
> 
> I tested the theory by forcing CMake to go through the fail path on my x86 system, resulting in the FDIV_SOURCE_PATH being empty and saw the same cmake errors in the build.
> 
> Currently, I have my code patched with the change I listed above and would be willing to put up a PR for the fix, but would need a little guidance as to whether there are other code paths which would need to be qualified to fully turn off this feature in the cases it is not supported or the user unselects it.

@rjknight Why not submit a PR?

**Comment by dncliss** [comment]
The PR linked to this issue does fix the problem, but we're still waiting on it to get merged into the 'master' branch upstream.

I'm trying to get an idea of when that might be (as per my question in that PR last week), but as of now we don't have any good feel on the timeframe.  So....timing TBD.  CI builds on ppc64le will remain broken in the interim.

**Comment by xdever** [comment]
I just run in the same issue on the newly released PyTorch 1.5.0 stable.

**Comment by dncliss** [comment]
> I just run in the same issue on the newly released PyTorch 1.5.0 stable.

You're building for ppc64le too?  That makes sense; unfortunately the proposed patch wasn't included in the recent 1.5.0 stable branch, so it will need to be applied during build-time when building 1.5.0.

**Comment by pearu** [comment]
The PR #35607 breaks build with the following error:
```
[1252/1634] Linking CXX executable bin/conv_to_nnpack_transform_test
FAILED: bin/conv_to_nnpack_transform_test 
: && /home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/x86_64-conda_cos6-linux-gnu-c++  -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/pearu/miniconda3/envs/pytorch-cuda-dev/include -L/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib -D__STDC_FORMAT_MACROS -Wno-deprecated -fvisibility-inlines-hidden -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG  -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib -Wl,-rpath-link,/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib -L/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib -rdynamic caffe2/CMakeFiles/conv_to_nnpack_transform_test.dir/transforms/conv_to_nnpack_transform_test.cc.o  -o bin/conv_to_nnpack_transform_test  -Wl,-rpath,/home/pearu/git/pytorch/pytorch/build/lib: lib/libgtest_main.a -Wl,--no-as-needed,/home/pearu/git/pytorch/pytorch/build/lib/libtorch.so -Wl,--as-needed -Wl,--no-as-needed,/home/pearu/git/pytorch/pytorch/build/lib/libtorch_cpu.so -Wl,--as-needed lib/libprotobuf.a lib/libc10.so -lmkl_intel_lp64 -lmkl_gnu_thread -lmkl_core /home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/libiomp5.so -lpthread /home/pearu/miniconda3/envs/pytorch-cuda-dev/x86_64-conda_cos6-linux-gnu/sysroot/usr/lib/libm.so /home/pearu/miniconda3/envs/pytorch-cuda-dev/x86_64-conda_cos6-linux-gnu/sysroot/usr/lib/libdl.so lib/libgtest.a -pthread && :
/home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/pearu/git/pytorch/pytorch/build/lib/libtorch_cpu.so: undefined reference to `pthreadpool_compute_1d_tiled'
/home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/pearu/git/pytorch/pytorch/build/lib/libtorch_cpu.so: undefined reference to `pthreadpool_compute_3d_tiled'
/home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/pearu/git/pytorch/pytorch/build/lib/libtorch_cpu.so: undefined reference to `caffe2::mobile_pthreadpool()'
/home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/pearu/git/pytorch/pytorch/build/lib/libtorch_cpu.so: undefined reference to `pthreadpool_compute_1d'
/home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/pearu/git/pytorch/pytorch/build/lib/libtorch_cpu.so: undefined reference to `pthreadpool_compute_2d'
/home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/pearu/git/pytorch/pytorch/build/lib/libtorch_cpu.so: undefined reference to `pthreadpool_compute_4d_tiled'
collect2: error: ld returned 1 exit status
```
in Ubuntu 18.04 when building from conda environment.

**Comment by eellison** [comment]
I also ran into the issue @pearu described.

**Comment by rjknight** [comment]
I'll take a look and see if I can resolve it.

**Comment by adamjstewart** [comment]
I'm encountering the same problem as initially reported in this issue, but on macOS x86_64, not ppc64le. I think this issue occurs any time you try to disable XNNPACK, regardless of architecture. I'll see if #35607 fixes the issue for me.

**Comment by rjknight** [comment]
I pushed up another PR which I think will resolve the issues @pearu  and @eellison were seeing
